File: complex objects/ios/ComplexRust/Rust/List/ListManager.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation

class ListManager: RustObject {
    let raw: OpaquePointer

    required init(raw: OpaquePointer) {
        self.raw = raw
    }

    func allCategories() -> [Category] {
        let categories = get_all_categories(self.raw)
        var allCategories: [Category] = []
        for index in 0..<category_list_count(categories) {
            let category = Category(raw: category_list_item_at(categories, index))
            allCategories.append(category)
        }
        return allCategories
    }

    func createCategory(withName name: String) -> Category {
        return Category(raw: category_new(self.raw, name))
    }

    func add(item: Item, toCategory category: Category) {
        category_manager_create_item(raw, item.raw, category.id)
    }

    func update(item: Item) throws {
        category_manager_update_item(raw, item.raw)
    }
}

File: complex/android/Toodle/app/src/main/java/com/mozilla/toodle/rust/ListManager.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package com.mozilla.toodle.rust;

public class ListManager extends RustObject {
    /* package-private */ ListManager(Toodle toodle) {
        rawPointer = JNA.INSTANCE.toodle_list(toodle.rawPointer);
    }

    public void createItem(String name, long dueDate) {
        JNA.INSTANCE.a_list_manager_create_item(
                rawPointer,
                name,
                dueDate
        );
    }

    @Override
    public void close() {
        JNA.INSTANCE.toodle_list_destroy(rawPointer);
    }
}